Shut off MD module upgrades after 7.4 EOL

andystaples opened the pull request:
Will need to be added as part of documentation work for #943 
Backport to 7.2 required, will link to that PR once this one is merged and rebased

### Pull request checklist

* [ ] My changes **do not** require documentation changes
  * [x] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
  * [x] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
  * [x] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

andystaples committed:
Shut off MD module upgrades after 7.4 EOL

andystaples committed:
Missed edit to one file checking date

AnatoliB commented on test/Unit/DependencyManagement/DependencyManagerTests.cs:
```diff
@@ -229,40 +229,49 @@ public void StartDependencyInstallationIfNeeded_InstallsSnapshotInForeground_Whe
-            _mockStorage.Setup(_ => _.GetDependencies()).Returns(GetAnyNonEmptyDependencyManifestEntries());
+            try
+            {
+                Environment.SetEnvironmentVariable("MDEnableAutomaticUpgrades", "true");
```
If we do just this, we will not have a test that verifies this behavior when the `MDEnableAutomaticUpgrades` variable is _not_ present (which is actually the most common case).

I don't have a simple solution, let's think about it. Ideally, we need to test all combinations of `MDEnableAutomaticUpgrades` (true, false, absent), both before the EOL moment and after the EOL moment, which makes it 6 test cases. But then we'll have to pass both `MDEnableAutomaticUpgrades` and a fixed "current time" to the tests. Which is doable, but perhaps there is an easier way.

andystaples replied:
I think there are more tests that are needed here, but wanted to get your thoughts on which ones are relevant. The test I added confirms that, if the Func<bool> passed to BackgroundDependencySnapshotMaintainer returns false, it does not attempt to install dependencies and instead warns to the logger. Any other cases you'd like me to add specific checks for?

AnatoliB replied:
Thank you, I think the tests we have now are adequate.

AnatoliB commented on src/DependencyManagement/BackgroundDependencySnapshotMaintainer.cs:
```diff
@@ -42,6 +46,16 @@ public void Start(string currentSnapshotPath, DependencyManifestEntry[] dependen
 
             _purger.SetCurrentlyUsedSnapshot(currentSnapshotPath, logger);
 
+            if (WorkerEnvironment.IsPowerShellSDKDeprecated() && !EnableAutomaticUpgrades)
```
This change will probably break the BackgroundDependencySnapshotMaintainer unit tests after the EOL date.

AnatoliB replied:
Also, this implementation will make the decision just once, on a worker start. Imagine a worker running for days, weeks, or months: it will already have the `_installAndPurgeTimer` running, and nothing is going to prevent it from installing modules on the next opportunity (if I understand this code correctly). We need to find a way to make this decision more dynamic, perhaps move the check closer to the place where new modules are being installed.

AnatoliB replied:
I think the beginning of `InstallAndPurgeSnapshots` is a better place for this check, and we can even throw an exception and let the existing catch block log it as a DependenciesUpgradeSkippedMessage warning. On the EOL date, the behavior will change without restarting the worker. I like it.

In this case, you also don't need to change `DependencyManager` or fix its unit tests anymore, as the `BackgroundDependencySnapshotMaintainer` is mocked there.

You will still need to take care of `BackgroundDependencySnapshotMaintainer` unit tests though: either provide all combinations of `MDEnableAutomaticUpgrades` and the current time, which still looks ugly. Or, much better, just add `Func<bool>` as a parameter to the `BackgroundDependencySnapshotMaintainer` constructor, so you can pass `() => WorkerEnvironment.IsPowerShellSDKDeprecated() && !EnableAutomaticUpgrades` when it's instantiated for production purposes, and `() => true` or `() => false` for test purposes. If you do this, you will have only two test cases for `BackgroundDependencySnapshotMaintainer`: one with the current behavior, and another one with the "blocked" behavior. I think we can skip testing the `WorkerEnvironment.IsPowerShellSDKDeprecated() && !EnableAutomaticUpgrades` expression itself.

andystaples replied:
Moved to InstallAndPurgeSnapshots and made the Func<bool> change

andystaples committed:
Address PR concerns
- Move ShouldEnableManagedDpendencyUpgrades logic to dedicated function
- Move check to InstallAndPurgeSnapshots for long-running workers
- Adjust tests to mock ShouldEnableManagedDpendencyUpgrades correctly

andystaples committed:
Revert InstallFunctionAppDependencies change

andystaples committed:
Revert DependencyManagerTests change

andystaples committed:
Fix test

AnatoliB commented on src/DependencyManagement/WorkerEnvironment.cs:
```diff
@@ -16,6 +16,8 @@ internal static class WorkerEnvironment
         private const string ContainerName = "CONTAINER_NAME";
         private const string LegionServiceHost = "LEGION_SERVICE_HOST";
 
+        private static readonly DateTime PowerShellSDKDeprecationDate = new DateTime(2026, 11, 10);
```
We'll need to remember to keep updating this in the future. Perhaps we should add a reminder item to https://github.com/Azure/azure-functions-powershell-worker/blob/dev/upgrade_ps_sdk_instructions.md

andystaples replied:
Added create_new_worker_instructions.md to document steps that must be taken when updating to a new PS 7 minor version (like 7.6). 
This file includes a note to update the Managed Dependencies shutoff date. It is likely we will find many new checklist items for this file as we start the 7.6 worker next year.

AnatoliB reviewed (approved):
LGTM

andystaples committed:
Add instructions for new PowerShell versions
- Managed dependencies shutoff date must be updated
